src/manifolds/Tucker.jl

```diff
@@ -0,0 +1,812 @@
+ambient space is represented as a full tensor.
+"""
+function inner(::Tucker, 𝔄::TuckerPoint, x::TuckerTVector, y::TuckerTVector)
+    ℭ = 𝔄.hosvd.core
+    dotprod = dot(x.Ċ, y.Ċ)
+    for d in 1:ndims(𝔄)
+        dotprod += dot(x.U̇[d] * tensor_unfold(ℭ, d), y.U̇[d] * tensor_unfold(ℭ, d))
+    end
+    return dotprod
+end
+inner(M::Tucker, 𝔄::TuckerPoint, x::TuckerTVector, y) = dot(embed(M, 𝔄, x), y)
+inner(M::Tucker, 𝔄::TuckerPoint, x, y::TuckerTVector) = dot(x, embed(M, 𝔄, y))
+
+"""
+    inverse_retract(ℳ::Tucker, A::TuckerPoint, B::TuckerPoint, r::ProjectionInverseRetraction)
+
+The projection inverse retraction on the Tucker manifold interprets `B` as a point in the
+ambient Euclidean space and projects it onto the tangent space at to `ℳ` at `A`.
+"""
+inverse_retract(
+    ::Tucker, ::TuckerPoint, ::TuckerPoint, ::ProjectionInverseRetraction
+)
+
+function inverse_retract!(
+    ℳ::Tucker, X, 𝔄::TuckerPoint, 𝔅::TuckerPoint, ::ProjectionInverseRetraction
+)
+    diffVector = embed(ℳ, 𝔅) - embed(ℳ, 𝔄)
+    return project!(ℳ, X, 𝔄, diffVector)
+end
+
+function isapprox(p::TuckerPoint, q::TuckerPoint; kwargs...)
+    ℳ = Tucker(size(p), size(p.hosvd.core))
+    return isapprox(embed(ℳ, p), embed(ℳ, q); kwargs...)
+end
+isapprox(::Tucker, p::TuckerPoint, q::TuckerPoint; kwargs...) = isapprox(p, q; kwargs...)
+function isapprox(M::Tucker, p::TuckerPoint, x::TuckerTVector, y::TuckerTVector; kwargs...)
+    return isapprox(embed(M, p, x), embed(M, p, y); kwargs...)
+end
+
+#=
+Determines whether there are tensors of dimensions n⃗ with multilinear rank r⃗
+=#
+function is_valid_mlrank(n⃗, r⃗)
+    return all(r⃗ .≥ 1) &&
+           all(r⃗ .≤ n⃗) &&
+           all(ntuple(i -> r⃗[i] ≤ prod(r⃗) ÷ r⃗[i], length(r⃗)))
+end
+
+@doc raw"""
+    manifold_dimension(::Tucker)
+
+The dimension of the manifold of $N_1 \times \dots \times N_D$ tensors of multilinear
+rank $R_1 \times \dots \times R_D$, i.e.
+```math
+    \mathrm{dim}(\mathcal{M}) = \prod_{d=1}^D R_d + \sum_{d=1}^D R_d (N_d - R_d).
+```
+"""
+manifold_dimension(::Tucker{n⃗,r⃗}) where {n⃗,r⃗} = prod(r⃗) + sum(r⃗ .* (n⃗ .- r⃗))
+
+@doc raw"""
+    Base.ndims(A :: TuckerPoint{T, D})
+
+The order of the tensor corresponding to the [`TuckerPoint`](@ref) `A`
+"""
+Base.ndims(::TuckerPoint{T,D}) where {T,D} = D
+
+number_eltype(::TuckerPoint{T,D}) where {T,D} = T
+number_eltype(::TuckerTVector{T,D}) where {T,D} = T
+
+"""
+    project(ℳ::Tucker, 𝔄::TuckerPoint, X)
+
+The least-squares projection of a tensor `X` to the tangent space to `ℳ` at `A`.
+"""
+project(::Tucker, ::Any, ::TuckerPoint, ::Any)
+
+function project!(ℳ::Tucker, Y, 𝔄::TuckerPoint, X)
+    ℬ = get_basis(ℳ, 𝔄, DefaultOrthonormalBasis())
+    coords = Vector{number_eltype(𝔄)}(undef, manifold_dimension(ℳ))
+    f!(i, ℬᵢ) = setindex!(coords, inner(ℳ, 𝔄, ℬᵢ, X), i)
+    foreach(f!, ℳ, 𝔄, ℬ)
+    return get_vector!(ℳ, Y, 𝔄, coords, ℬ)
+end
+
+representation_size(::Tucker{N}) where {N} = N
```

Comment:
This is incorrect, when points and vectors on a manifold are represented using special types, `representation_size` should return `nothing` (which is the default). Did you need it to make some test pass?